readthedocs/organizations/views/decorators.py

```diff
@@ -0,0 +1,105 @@
+import functools
+
+from django.conf import settings
+from django.http import Http404, HttpResponseRedirect
+from django.shortcuts import resolve_url
+from django.urls import NoReverseMatch
+from django.utils.http import urlencode
+from structlog import get_logger
+
+from readthedocs.organizations.models import Organization
+
+log = get_logger(__name__)
+
+
+def redirect_if_organization_unspecified(dispatch):
```

Comment:
What is the purpose of these two decorators? On #938 we discussed just adding a single view for an organization selection (and later a project selection) page. It doesn't seem necessary to decorate any other views or wrap them with this logic.

We don't need this on all of our URLs, only specific ones that we link to commonly, like the subscription edit URL. So, to me, this approach seems like it adds more complexity than problems it is solving.

Is there some other reason that this was implemented?

Reply:
With a decorator, you can decorate the `dispatch` method on exactly the CBVs that requires the functionality.

With a mixin, you need to somehow understand whateever dispatch functions that are used in the class hierarchy and modify those. We already have a mixin that modifies the `dispatch` method :exploding_head: 

This is where view decorators are perfect.

Reply:
Also, you're more than welcome to suggest exactly the views that should have this functionality, then I can add the decorator on exactly those views :100:

Reply:
> With a decorator, you can decorate the dispatch method on exactly the CBVs that requires the functionality.

I wasn't asking about the code mechanics here of a mixin vs decorator, I was asking more about why we're adding this feature to every view.

> Also, you're more than welcome to suggest exactly the views that should have this functionality

Well, what I described above is that I feel we don't need this pattern at all. We only talked about implementing the redirect view, and this is enough for our use case.

Reply:
> I wasn't asking about the code mechanics here of a mixin vs decorator, I was asking more about why we're adding this feature to every view.

I looked at the views and their URL patterns and it seemed to be possible to add to all the views. It seems super nice that we can now create a URL for any organization-based settings page when we need to do support and point people to a specific page?

Reply:
That was the intention with the redirect view, so I'm not sure we need two different ways of doing this. The redirect view added here is explicit and isolated and doesn't involve altering all our views, so I think dialing this work back is still the correct call.

Reply:
> With a mixin, you need to somehow understand whateever dispatch functions that are used in the class hierarchy and modify those. We already have a mixin that modifies the dispatch method 🤯

This is my point. We already have an established pattern in the codebase of using mixins to achieve this result. Adding a new way of doing it is bad, regardless of whether there is a small technical benefit to doing it. My thought is that we should just implement the new redirect logic as a mixin.

The question around whether to implement this functionality as a mixin vs. a dedicated view is another question entirely. I tend to agree this implementation adds a lot of complexity compared to the redirect view we discussed in the issue, and was my understanding of what we were going to implement.

Reply:
I think that the URL rewriting that's required in order to inject slugs of organizations has been underestimated in the "let's just have a redirect page" scheme. This change includes the redirect page but leverages existing views with slugs to generate the redirect page's parameters.

Reply:
In the example with `?view=organization_edit`, this requires someone wanting to share a URL to look up the exact view name in the source code's Django urlconf, which I thought was very very nice to avoid.

Reply:
But yes, we can peel off the layer with the decorator function and write the view name manually for the redirect page. I just thought it'd be nice to not have to do that :)

Reply:
That also means sharing URLs directly to the redirect page rather than the more convenient URLs with just `-` as slug.

Reply:
I think this is a case where we value complexity/usability ratio. Considering this was originally thought as an internal view that we will only use while doing support, I'm fine reducing the complexity as much as possible at expenses of keeping the usability a little more complex (e.g. we need to know the internal Django view name to put in the URL query string)

I'm fine considering an easier to use and more complex implementation in the future if we find it's required or we find that the initial redirect view is not covering all the use cases; but I think we are going to far here for the initial implementation.

Even if this implementation would have been better than the one discussed in the original issue, it makes me some noise that something different than what we agreed on was implemented. I'd like to have discussed the new idea in advance before implementing it and having this discussion after that.